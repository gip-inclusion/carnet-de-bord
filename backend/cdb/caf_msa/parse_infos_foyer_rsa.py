from typing import List

import lxml.etree as etree
from pydantic import BaseModel


class Personne(BaseModel):
    nir: str
    soumis_droit_et_devoir: str


class CafMsaInfosFoyer(BaseModel):
    matricule: str
    etat_droit_rsa: str
    personnes: List[Personne]
    motif_versement_rsa: str | None
    motif_cloture_rsa: str | None
    date_cloture_rsa: str | None
    sans_domicile_fixe: bool


def parse_infos_foyer_rsa(node: etree._ElementTree):

    personneNodes: list[etree._ElementTree] = node.findall("Personne", namespaces=None)

    personnes = [
        Personne(
            nir=personne.findtext("Identification/NIR", default=None, namespaces=None),
            soumis_droit_et_devoir=personne.findtext(
                "MontantCalculDroitRSAPersonne/TOPPERSDRODEVORSA",
                default=None,
                namespaces=None,
            ),
        )
        for personne in personneNodes
    ]

    return CafMsaInfosFoyer(
        matricule=node.findtext(
            "IdentificationRSA/Organisme/MATRICULE", default=None, namespaces=None
        ),
        personnes=personnes,
        etat_droit_rsa=node.findtext(
            "PrestationRSA/SituationDossierRSA/EtatDossierRSA/ETATDOSRSA",
            default=None,
            namespaces=None,
        ),
        motif_versement_rsa=node.findtext(
            "PrestationRSA/SituationDossierRSA/SuspensionVersement/MOTISUSVERSRSA",
            default=None,
            namespaces=None,
        ),
        motif_cloture_rsa=node.findtext(
            "PrestationRSA/SituationDossierRSA/FinDroit/MOTICLORSA",
            default=None,
            namespaces=None,
        ),
        date_cloture_rsa=node.findtext(
            "PrestationRSA/SituationDossierRSA/FinDroit/DTCLORSA",
            default=None,
            namespaces=None,
        ),
        sans_domicile_fixe=node.findtext(
            "PrestationRSA/DetailDroitRSA/TroncCommunDroitRSA/TOPSANSDOMFIXE",
            default=None,
            namespaces=None,
        ),
    )
