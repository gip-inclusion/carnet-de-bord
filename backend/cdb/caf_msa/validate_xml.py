import io
import mmap
import os
from typing import Any, Generator

from lxml import etree

current_dir = os.path.dirname(os.path.realpath(__file__))
parser = etree.XMLParser(remove_blank_text=True)


def validate_xml(data: io.BufferedReader) -> bool:
    xmlschema_doc = etree.parse(
        os.path.join(current_dir, "vrsb0801.xsd"), parser=parser
    )
    xmlschema = etree.XMLSchema(xmlschema_doc)
    xml_doc = etree.parse(data, parser=parser)
    xmlschema.assertValid(xml_doc)
    return True


class XMLTagStream:
    """
    Shamelessly copy from
    https://realpython.com/python-xml-parser/#explore-third-party-xml-parser-libraries
    It will extract an xml node from the xml file

    """

    def __init__(self, file: io.BufferedReader, tag_name, encoding="utf-8"):
        self.file = file
        self.stream = mmap.mmap(self.file.fileno(), 0, access=mmap.ACCESS_READ)
        self.tag_name = tag_name
        self.encoding = encoding
        self.start_tag = f"<{tag_name}>".encode(encoding)
        self.end_tag = f"</{tag_name}>".encode(encoding)

    def __enter__(self):
        return self

    def __exit__(self, *args, **kwargs):
        self.stream.close()

    def __iter__(self) -> Generator[etree._ElementTree, Any, Any]:
        end = 0
        while (begin := self.stream.find(self.start_tag, end)) != -1:
            end = self.stream.find(self.end_tag, begin)
            yield self.parse(self.stream[begin : end + len(self.end_tag)])

    def parse(self, chunk) -> etree._ElementTree:
        return etree.parse(io.StringIO(chunk.decode(self.encoding)), parser=parser)
