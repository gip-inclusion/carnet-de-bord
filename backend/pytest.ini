[pytest]
asyncio_mode = auto
testpaths = "tests"
addopts = --strict-markers --strict-config
markers =
    graphql: Used for tests that rely on GraphQL. Those tests actually write to the database, which needs to be re-seeded. Seeding being expensive, prefer the use of SQL.
#log_cli = true
#log_level = INFO
