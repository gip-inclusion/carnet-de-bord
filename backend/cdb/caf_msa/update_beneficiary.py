import hashlib
import io
import logging
from datetime import date
from typing import List, Tuple
from uuid import UUID

from cdb.api.core.emails import notify_manager_after_cafmsa_import
from cdb.api.core.exceptions import FindResultException
from cdb.api.core.settings import settings
from cdb.api.db.graphql.beneficiary import get_beneficiary_by_nir, update_beneficiary
from cdb.api.db.graphql.get_client import gql_client
from cdb.api.db.graphql.manager import get_manager_by_account_id
from cdb.caf_msa.parse_infos_foyer_rsa import (
    CafInfoFlux,
    CafMsaInfosFoyer,
    parse_caf_file,
    transform_cafMsaFoyer_to_beneficiary,
)


async def update_beneficiaries_from_cafmsa(
    account_id: UUID, jwt_token: str, file: io.BufferedReader
) -> None:
    data: Tuple[CafInfoFlux, List[CafMsaInfosFoyer]] = parse_caf_file(file)
    infos, foyers = data
    count = 0
    count_error = 0
    count_success = 0
    async with await gql_client(
        url=settings.graphql_api_url,
        headers={
            "x-hasura-admin-secret": settings.hasura_graphql_admin_secret,
            "x-hasura-use-backend-only-permissions": "true",
            "Authorization": "Bearer " + jwt_token,
        },
    ) as gql_session:
        for foyer in foyers:
            for personne in foyer.personnes:
                count += 1
                try:
                    beneficiary = await get_beneficiary_by_nir(
                        gql_session, personne.nir
                    )
                    if beneficiary:
                        beneficiary_update = transform_cafMsaFoyer_to_beneficiary(
                            personne, foyer
                        )
                        external_data_to_save = {
                            "flux": infos.dict(),
                            "foyer": foyer.dict(),
                        }
                        sha = hashlib.sha256(str(data).encode()).hexdigest()
                        try:
                            await update_beneficiary(
                                gql_session,
                                beneficiary.id,
                                beneficiary_update,
                                sha,
                                external_data_to_save,
                            )
                            count_success += 1
                        except Exception as error:
                            count_error += 1
                            logging.error(
                                "échec de la mise à jour du bénéficiaire %s "
                                "caf/msa; %s",
                                beneficiary.id,
                                error,
                            )

                except Exception as error:
                    count_error += 1
                    logging.error(
                        "echec lors de la récupération du bénéficiaire "
                        "depuis le nir %s",
                        error,
                    )

        manager_email = await get_manager_by_account_id(gql_session, account_id)
        if manager_email:
            notify_manager_after_cafmsa_import(
                manager_email.email,
                date.today().strftime("%d/%m/%Y"),
                count,
                count_success,
                count_error,
            )
        else:
            raise FindResultException
