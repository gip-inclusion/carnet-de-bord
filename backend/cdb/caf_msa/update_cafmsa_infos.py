import hashlib
import logging
import time
from datetime import date, timedelta
from tempfile import SpooledTemporaryFile
from uuid import UUID

from gql.client import AsyncClientSession

from cdb.api.core.emails import notify_manager_after_cafmsa_import
from cdb.api.db.graphql.beneficiary import (
    BeneficiaryRsaInfos,
    get_beneficiary_by_nir,
    update_beneficiary,
)
from cdb.api.db.graphql.get_client import gql_client_backend_only
from cdb.api.db.graphql.manager import get_manager_by_account_id
from cdb.caf_msa.parse_infos_foyer_rsa import (
    CafBeneficiary,
    CafInfoFlux,
    CafMsaInfosFoyer,
    parse_caf_file,
    transform_cafMsaFoyer_to_beneficiary,
)


async def update_cafmsa_for_beneficiaries(
    account_id: UUID, authorization: str, file: SpooledTemporaryFile
) -> None:
    count = 0
    count_error = 0
    count_success = 0
    time.time()

    infos: CafInfoFlux | None = None

    async with gql_client_backend_only(bearer_token=authorization) as session:
        start_time = time.time()

        parsed = parse_caf_file(file)
        maybe_infos = next(parsed)
        if not isinstance(maybe_infos, CafInfoFlux):
            raise CafXMLMissingNodeException

        infos = maybe_infos

        for node in parsed:
            if not isinstance(node, CafMsaInfosFoyer) or not infos:
                raise CafXMLMissingNodeException
            for beneficiary in node.beneficiaries:
                try:
                    result = await save_cafmsa_infos_for_beneficiary(
                        gql_session=session,
                        flux_info=infos,
                        foyer=node,
                        caf_beneficiary=beneficiary,
                    )
                    if result:
                        count_success += 1
                except Exception as error:
                    logging.error(
                        "Erreur lors du traitement du dossier %s",
                        error,
                    )
                    count_error += 1
                finally:
                    count += 1

        logging.info(
            "Mise à jour de %s/%s dossiers reçus (%s erreurs) en %s",
            count_success,
            count,
            count_error,
            timedelta(seconds=time.time() - start_time),
        )

        manager = await get_manager_by_account_id(session, account_id)
        if manager:
            logging.info(
                "Notification de la fin du traitement au manager %s",
                manager.email,
            )
            notify_manager_after_cafmsa_import(
                manager.email,
                date.today().strftime("%d/%m/%Y"),
                count,
                count_success,
                count_error,
            )
        else:
            logging.warn(
                "manager (id=%s) introuvable",
                account_id,
            )


async def save_cafmsa_infos_for_beneficiary(
    gql_session: AsyncClientSession,
    flux_info: CafInfoFlux,
    foyer: CafMsaInfosFoyer,
    caf_beneficiary: CafBeneficiary,
) -> BeneficiaryRsaInfos | None:
    beneficiary = await get_beneficiary_by_nir(gql_session, caf_beneficiary.nir)

    if beneficiary:
        beneficiary_update = transform_cafMsaFoyer_to_beneficiary(
            caf_beneficiary, foyer
        )
        external_data_to_save = {
            "flux": flux_info.dict(),
            "foyer": foyer.dict(),
        }
        sha = hashlib.sha256(str(external_data_to_save).encode()).hexdigest()

        return await update_beneficiary(
            gql_session,
            beneficiary.id,
            beneficiary_update,
            sha,
            external_data_to_save,
        )

    else:
        logging.info("Bénéficiaire non trouvé.", caf_beneficiary.nir)


class CafXMLMissingNodeException(Exception):
    """
    utility class when some node in xml are missing
    """
