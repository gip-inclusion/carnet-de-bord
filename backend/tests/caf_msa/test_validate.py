import io

import lxml.etree as etree
import pytest

from cdb.caf_msa.parse_infos_foyer_rsa import parse_infos_foyer_rsa
from cdb.caf_msa.validate_xml import XMLTagStream, validate_xml


async def test_validate_incorrect_file(flux_quotidien_caf_invalid: io.BufferedReader):
    with pytest.raises(etree.DocumentInvalid):
        validate_xml(flux_quotidien_caf_invalid)


async def test_validate_correct_file(flux_mensuel_caf: io.BufferedReader):
    assert validate_xml(flux_mensuel_caf)


async def test_extract_data_using_xml_tag_stream(flux_mensuel_caf: io.BufferedReader):
    foyers = []
    with XMLTagStream(flux_mensuel_caf, "InfosFoyerRSA") as stream:
        for infosFoyer in stream:
            try:
                foyers.append(parse_infos_foyer_rsa(infosFoyer))
            except Exception as error:
                print(error)
    assert len(foyers) == 2
