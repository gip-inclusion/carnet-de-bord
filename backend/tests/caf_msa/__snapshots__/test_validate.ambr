# name: test_extract_data_using_xml_tag_stream
  list([
    CafMsaInfosFoyer(matricule='205599000000000', etat_droit_rsa='5', personnes=[Personne(nir='2820251108030', soumis_droit_et_devoir='1'), Personne(nir='1770393050005', soumis_droit_et_devoir='0')], motif_versement_rsa=None, motif_cloture_rsa='PCG', date_cloture_rsa='2021-11-30', sans_domicile_fixe=False),
    CafMsaInfosFoyer(matricule='208956400000000', etat_droit_rsa='2', personnes=[Personne(nir='2730799404078', soumis_droit_et_devoir='1')], motif_versement_rsa=None, motif_cloture_rsa='PCG', date_cloture_rsa='2021-11-30', sans_domicile_fixe=False),
  ])
# ---
